File: apps/mobile/ios/Runner/PaddleOCR/ReceiptDocumentOrientation.swift
import Foundation

enum ReceiptDocumentOrientation: Equatable {
  case upright
  case clockwise90
  case upsideDown
  case counterclockwise90

  func transform(
    _ point: SettleoraOcrPoint,
    sourceWidth: Int,
    sourceHeight: Int
  ) -> SettleoraOcrPoint {
    let maxX = Double(max(0, sourceWidth - 1))
    let maxY = Double(max(0, sourceHeight - 1))
    switch self {
    case .upright:
      return point
    case .clockwise90:
      return SettleoraOcrPoint(
        x: max(0, min(maxY, maxY - point.y)),
        y: max(0, min(maxX, point.x))
      )
    case .upsideDown:
      return SettleoraOcrPoint(
        x: max(0, min(maxX, maxX - point.x)),
        y: max(0, min(maxY, maxY - point.y))
      )
    case .counterclockwise90:
      return SettleoraOcrPoint(
        x: max(0, min(maxY, point.y)),
        y: max(0, min(maxX, maxX - point.x))
      )
    }
  }

  static func select(
    lineDimensions: [(width: Double, height: Double)],
    reverseRecognition: Bool
  ) -> ReceiptDocumentOrientation {
    let verticalLines = lineDimensions.filter {
      max(1, $0.height) / max(1, $0.width) >= 1.5
    }.count
    let sideways = verticalLines > lineDimensions.count - verticalLines
    switch (sideways, reverseRecognition) {
    case (true, true): return .clockwise90
    case (true, false): return .counterclockwise90
    case (false, true): return .upsideDown
    case (false, false): return .upright
    }
  }
}

File: apps/mobile/android/app/src/main/kotlin/com/example/mobile/ocr/ReceiptDocumentOrientation.kt
package com.example.mobile.ocr

/** The document-space transform needed to make detected receipt rows upright. */
internal enum class ReceiptDocumentOrientation {
    UPRIGHT,
    CLOCKWISE_90,
    UPSIDE_DOWN,
    COUNTERCLOCKWISE_90;

    fun transform(point: SettleoraOcrPoint, sourceWidth: Int, sourceHeight: Int): SettleoraOcrPoint =
        when (this) {
            UPRIGHT -> point
            CLOCKWISE_90 -> SettleoraOcrPoint(
                (sourceHeight - 1f - point.y).coerceIn(0f, sourceHeight - 1f),
                point.x.coerceIn(0f, sourceWidth - 1f),
            )
            UPSIDE_DOWN -> SettleoraOcrPoint(
                (sourceWidth - 1f - point.x).coerceIn(0f, sourceWidth - 1f),
                (sourceHeight - 1f - point.y).coerceIn(0f, sourceHeight - 1f),
            )
            COUNTERCLOCKWISE_90 -> SettleoraOcrPoint(
                point.y.coerceIn(0f, sourceHeight - 1f),
                (sourceWidth - 1f - point.x).coerceIn(0f, sourceWidth - 1f),
            )
        }

    companion object {
        private const val VERTICAL_LINE_RATIO = 1.5f

        fun select(
            lineDimensions: List<Pair<Float, Float>>,
            reverseRecognition: Boolean,
        ): ReceiptDocumentOrientation {
            val verticalLines = lineDimensions.count { (rawWidth, rawHeight) ->
                val width = rawWidth.coerceAtLeast(1f)
                val height = rawHeight.coerceAtLeast(1f)
                height / width >= VERTICAL_LINE_RATIO
            }
            val horizontalLines = lineDimensions.size - verticalLines
            val sideways = verticalLines > horizontalLines
            return when {
                sideways && reverseRecognition -> CLOCKWISE_90
                sideways -> COUNTERCLOCKWISE_90
                reverseRecognition -> UPSIDE_DOWN
                else -> UPRIGHT
            }
        }
    }
}
